### ISSUE TITLE
Support for correlated subquery referencing csv/parquet files

### ISSUE THREAD

**Issue by chrisbrain** [duckdb/duckdb#4920]
Not sure if this is an issue with my SQL/understanding or whether it's not supported but is it possible for a correlated subquery to work against parquet files?

I'll use a CSV file here as an example for clarity but the idea would be identical. Say I have the following file:

x.csv
```
id, name
1, a
2, b
```

And then the following related csvs:

y1.csv
```
id, val
1, 1
1, 2
```

y2.csv
```
id, val
2, 3
2, 4
```

What I am trying to get is a table showing:
```
id, name, sum(val)
1, a, 3
2, b, 7
```

I am thinking the query would look something like this:
```
select id, name, (select sum(val) from read_csv_auto('y' || parent.id || '.csv' || '')) from 'x.csv' as parent;
```

As you can see I am trying to concatenate the name of the relevant csv file for the current row being processed but I am getting the following error:

```
Error: Binder Error: Table function requires a constant parameter
```

And the same with the parquet equivalent.

I can see that concatenation does work in these functions as the following runs OK:
```
select * from read_csv_auto('y' || '1' || '.csv');
```

**Comment by Mytherin** [comment]
No, this is fundamentally not possible. During binding the system has to know the output types of every table function to be able to resolve the SQL statement, and the output columns of `read_csv_auto` depend on which file is being read. This means that the SQL statement cannot be resolved until we know which files are being read.

You could of course construct such a SQL statement outside of the system, or query all files and join them together based on the filename, e.g.:

```sql
SELECT *
FROM read_csv_auto('y*.csv', filename=True) AS y
JOIN 'x.csv' AS x 
ON (y.filename=concat('y', x.id, '.csv')
```

**Comment by chrisbrain** [comment]
Thanks for the quick reply @Mytherin.

I don't think the that workaround will work unfortunately as actually I am using httpfs (I was just trying to keep the example above minimal) which does not I don't think support the * syntax (globbing?).

Not sure if there is another workaround where I manually concatenate all the file names in the sql..

**Comment by Mytherin** [comment]
HTTPFS should support globbing as well.

**Comment by chrisbrain** [comment]
It says in the docs https://duckdb.org/docs/extensions/httpfs:

```
For pure HTTP(S), only file reading is supported. For object storage using the S3 API, the httpfs extension supports reading/writing/globbing files.
```

And at the moment I am actually writing my own httpfs server as I am generating the parquet files when they are requested so not even sure what the globbing API would look like I would have to support.

**Comment by Mause** [comment]
There's a difference between the HTTP Filesystem extension, and pure HTTP(S) servers. We support globbing via S3 and S3 compatible APIs, but there's no standard for listing files via pure HTTP(S) unfortunately

**Comment by Mytherin** [comment]
Ah yes, we do indeed only support globbing for S3 reads, not for HTTP reads. 

> Not sure if there is another workaround where I manually concatenate all the file names in the sql..

`read_parquet` can also take a list of files as a parameter, e.g.:

```sql
select * from read_parquet(['test.parquet', 'test2.parquet']) limit 5;
```

I just realized `read_csv_auto` does not support this yet, I will push a fix for that.

**Comment by chrisbrain** [comment]
OK thanks both! I guess I'll let you decide whether to close this before or after that fix goes in.

**Comment by Mytherin** [comment]
The CSV fix has been merged (#4922)
